modules/cloud-auth/google-auth.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2023 Attila Szakacs
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "google-auth.hpp"
+
+#include <exception>
+#include <fstream>
+
+#include "compat/cpp-start.h"
+#include "scratch-buffers.h"
+#include "compat/cpp-end.h"
+
+using namespace syslogng::cloud_auth::google;
+
+ServiceAccountAuthenticator::ServiceAccountAuthenticator(const char *key_path, const char *audience_)
+{
+  picojson::value key_json;
+
+  try
+    {
+      std::ifstream key_file(key_path);
+      std::string parse_error = picojson::parse(key_json, key_file);
+      key_file.close();
+
+      if (!parse_error.empty())
+        throw std::runtime_error(std::string("Failed to parse key file: ") + parse_error);
+    }
+  catch (const std::ifstream::failure &e)
+    {
+      throw std::runtime_error(std::string("Failed to open key file: ") + e.what());
+    }
+
+  try
+    {
+      email = key_json.get("client_email").get<std::string>();
+      private_key_id = key_json.get("private_key_id").get<std::string>();
+      private_key = key_json.get("private_key").get<std::string>();
+    }
+  catch (const std::runtime_error &e)
+    {
+      throw std::runtime_error(std::string("Failed to get the necessary fields from the key file: ") + e.what());
+    }
+
+  if (audience_ == nullptr)
+    throw std::runtime_error(std::string("audience() is mandatory"));
+  audience = audience_;
+}
+
+void
+ServiceAccountAuthenticator::handle_http_header_request(HttpHeaderRequestSignalData *data)
+{
+  /* Scratch Buffers are marked at this point in http-worker.c */
+  GString *buffer = scratch_buffers_alloc();
+  g_string_append(buffer, "Authorization: Bearer ");
+
+  std::chrono::system_clock::time_point issued_at = jwt::date::clock::now();
+  std::chrono::system_clock::time_point expires_at = issued_at + std::chrono::seconds{3600};
```

Comment:
Is this 3600 configurable? Do we plan to configure? Could you give a name for it?

Reply:
Yes, good idea. Fixed! :)